central/reprocessor/reprocessor.go

```diff
@@ -481,6 +497,165 @@ func (l *loopImpl) reprocessImagesAndResyncDeployments(fetchOpt imageEnricher.Fe
+		FetchOpt: fetchOpt,
+	}, image)
+
+	if err != nil {
+		log.Errorw("Error enriching image", logging.ImageName(image.GetName().GetFullName()), logging.ImageID(image.GetId()), logging.Err(err))
+		return nil, false
+	}
+	if result.ImageUpdated {
+		if err := l.risk.CalculateRiskAndUpsertImageV2(image); err != nil {
+			log.Errorw("Error upserting image into datastore",
+				logging.ImageName(image.GetName().GetFullName()), logging.ImageID(image.GetId()), logging.Err(err))
+			return nil, false
+		}
+		// We need to fetch the image again to make sure all fields are populated.
+		// GetImage will internally call a Merge function which will use the CVEEdges table to enrich fields like
+		// FirstImageOccurrence and FirstSystemOccurrence.
+		newImage, exists, err := l.imagesV2.GetImage(allAccessCtx, id)
+		if err != nil {
+			log.Errorw("Error fetching image from database", logging.ImageName(image.GetName().GetFullName()), logging.ImageID(image.GetId()), logging.Err(err))
+			return nil, false
+		}
+		if !exists {
+			log.Errorw("The image was not found after enrichement", logging.ImageName(image.GetName().GetFullName()), logging.ImageID(image.GetId()))
+			return nil, false
+		}
+		return newImage, true
+	}
+	return image, true
+}
+
+func (l *loopImpl) reprocessImagesV2AndResyncDeployments(fetchOpt imageEnricher.FetchOption,
+	imgReprocessingFunc imageReprocessingFuncV2, imageQuery *v1.Query) {
+	if l.stopSig.IsDone() {
+		return
+	}
+	results, err := l.imagesV2.Search(allAccessCtx, imageQuery)
+	if err != nil {
+		log.Errorw("Error searching for active image IDs", logging.Err(err))
+		return
+	}
+
+	log.Infof("Found %d images to scan", len(results))
+	if len(results) == 0 {
+		return
+	}
+
+	sema := semaphore.NewWeighted(imageReprocessorSemaphoreSize)
+	wg := concurrency.NewWaitGroup(0)
+	nReprocessed := atomic.NewInt32(0)
+	skipClusterIDs := maputil.NewSyncMap[string, struct{}]()
+	for _, result := range results {
+		wg.Add(1)
+		if err := sema.Acquire(concurrency.AsContext(&l.stopSig), 1); err != nil {
+			log.Errorw("Reprocessing stopped", logging.Err(err))
+			return
+		}
+		// Duplicates can exist if the image is within multiple deployments
+		clusterIDSet := set.NewStringSet(result.Matches[imageClusterIDFieldPath]...)
+		go func(id string, clusterIDs set.StringSet) {
+			defer sema.Release(1)
+			defer wg.Add(-1)
+
+			image, successfullyProcessed := l.reprocessImageV2(id, fetchOpt, imgReprocessingFunc)
+			if !successfullyProcessed {
+				return
+			}
+			nReprocessed.Inc()
+
+			utils.FilterSuppressedCVEsNoCloneV2(image)
+			utils.StripCVEDescriptionsNoCloneV2(image)
+
+			convertedImage := utils.ConvertToV1(image)
+			// Send the updated image to relevant clusters.
+			for clusterID := range clusterIDs {
+				conn := l.connManager.GetConnection(clusterID)
+				if conn == nil {
+					continue
+				}
+
+				msg := &central.MsgToSensor{
+					Msg: &central.MsgToSensor_UpdatedImage{
+						UpdatedImage: convertedImage,
+					},
+				}
```

Comment:
The image sent to sensor SHOULD contain all known names for the image - or we risk cache misses / outages for sensors that are not also updated.